### ISSUE TITLE
"there is no disk in the drive" with x64 build under Windows 10

### ISSUE THREAD

**Issue by mindplay-dk** [WPN-XM/WPN-XM#514]
With `WPNXM-0.8.6-Full-Setup-php56-w64` under Windows 10, I get an error prompt (twice) on launch, saying "there is no disk in the drive".

Wiped out my WPN-XM installation completely and started with a clean install, same thing.

I tracked it down to the CURL extension by disabling one extension at a time.

Tried replacing various files with ones from the official `php-5.6.16-nts-Win32-VC11-x64.zip`, and narrowed it down to `libeay32.dll` - replacing this file fixes the issue. (to make sure this is what fixed it, I restored all other files I had attempted to replace with their originals from your package.)

In summary, I don't think the `libeay32.dll` that comes with the latest release works on Win10-x64.
##

**Comment by mindplay-dk** [comment]
FYI, have been changing this file back and forth to confirm the issue comes back - it does.

**Comment by jakoch** [comment]
Hey Rasmus,

thank you for reporting this issue. 

Ok, i'm trying to reproduce the issue and these are my steps:

1. The installer is: `WPNXM-0.8.6-Full-Setup-php56-w64`
2. This installer uses the following package registry
    - https://github.com/WPN-XM/registry/blob/master/full-0.8.6-php5.6-w64.json
3. The package is "php-x64"
    - https://github.com/WPN-XM/registry/blob/master/full-0.8.6-php5.6-w64.json#L20
4. The download link is:
    - http://wpn-xm.org//get.php?s=php-x64&v=5.6.12
5. That downloads:
    - `php-5.6.12-nts-Win32-VC11-x64.zip`
    - from http://windows.php.net/downloads/releases/archives/php-5.6.12-nts-Win32-VC11-x64.zip
6. Extracting zip

Ok, now what exactly should i look for? The `curl` extension causes the problem?

8. Running `php -v` to list the PHP version:

    PHP 5.6.12 (cli) (built: Aug  6 2015 12:05:25)
    Copyright (c) 1997-2015 The PHP Group
    Zend Engine v2.6.0, Copyright (c) 1998-2015 Zend Technologies

9. Running `php -m` to list the PHP extensions loaded

```
    [PHP Modules]
    bcmath
    calendar
    Core
    ctype
    date
    dom
    ereg
    filter
    ftp
    hash
    iconv
    json
    libxml
    mcrypt
    mhash
    mysqlnd
    odbc
    pcre
    PDO
    Phar
    Reflection
    session
    SimpleXML
    SPL
    standard
    tokenizer
    wddx
    xml
    xmlreader
    xmlwriter
    zip
    zlib

    [Zend Modules]

```

Hmm, works here.

10. On top of that goes our modified PHP config for PHP 5.6.
 - https://github.com/WPN-XM/WPN-XM/blob/master/configs/php/php56.ini-dev
 - Renamed to `php.ini` dropped into PHP folder. 

11. `php -m` lists curl..

Hmm... 

It "could" be that one of the extensions installed on top is not "64", but "32" instead.
That would clash with the underlying PHP. But all are bitsize 64: https://github.com/WPN-XM/registry/blob/master/full-0.8.6-php5.6-w64.json#L21-L35

Disabling the extensions is a good idea step-wise is a good idea to find such an issue.

I will test the installer package now.

**Comment by jakoch** [comment]
Ok, i've tested the installer package on Windows-10-Pro-64.
I've started Nginx & PHP, then webinterface. It works out of the box.

Can't reproduce your issue.

> I get an error prompt (twice) on launch, saying "there is no disk in the drive".

Which further steps are needed to come close to this error?

**Comment by mindplay-dk** [comment]
A simple `php -v` was enough to reproduce the "no disk in drive" dialog on my system.

I did a little more digging, and it's a commonly-reported issue on various forums etc. so it must be system-dependent - it appears to be dependent on your drive-letter assignments somehow.

It's one of those issues that comes and goes - here's [one bug-report](https://bugs.php.net/bug.php?id=68312) which is currently closed, but has been reopened and closed several times the past couple of years, so I'm guessing it's one of those phantom bugs that are insanely difficult to pinpoint...

**Comment by jakoch** [comment]
Thank you for digging.

Interesting bug-report. I would say its a build/compile problem of the php extension curl.
They simply used a fixed path for the config during compilation.
After the compilation it becomes the hardcoded fallback, when some config vars are missing on the env.

> When none of OPENSSL_CONF or SSLEAY_CONF environmental var was found, that hardcoded path will be used in PHP as well. PHP should use a hardcoded path to c:\usr\local\ssl

Please try setting one of these constants as env var pointing to the \bin\openssl folder, e.g.

`set OPENSSL_CONF=c:/server/openssl/openssl.cfg`

**Comment by mindplay-dk** [comment]
I will try that in the morning.

It's odd that swapping out the binary would fix the problem though?

**Comment by mindplay-dk** [comment]
Well, this is weird. I restored all the files to their previous versions, but now there's no problem - the error doesn't occur anymore, even after a reboot.

The only possible explanation for that is, the first time it finds this stuff, it caches either the location or the actual result somewhere, right?

So I have no idea how to reproduce this now...

**Comment by mindplay-dk** [comment]
Okay, so it happened again this morning - this time it happens with `php-win.exe` only, which I discovered only because the WPN-XM control panel appears to run it on startup (?)

This time, the error was repeatable, and so the first thing I tried, was swapping out `libeay32.dll`, which fixed the problem immediately.

This confirms both of my suspicions, since, here's the output with the DLL that comes with WPN-XM:

```
$ php-win -r "print_r(openssl_get_cert_locations());"
Array
(
    [default_cert_file] => D:\tmp\openssl-1.0.1p\vc11\x64/cert.pem
    [default_cert_file_env] => SSL_CERT_FILE
    [default_cert_dir] => D:\tmp\openssl-1.0.1p\vc11\x64/certs
    [default_cert_dir_env] => SSL_CERT_DIR
    [default_private_dir] => D:\tmp\openssl-1.0.1p\vc11\x64/private
    [default_default_cert_area] => D:\tmp\openssl-1.0.1p\vc11\x64
    [ini_cafile] =>
    [ini_capath] =>
)
```

And here it is with the one from the official PHP distro:

```
$ php-win -r "print_r(openssl_get_cert_locations());"
Array
(
    [default_cert_file] => c:/usr/local/ssl/cert.pem
    [default_cert_file_env] => SSL_CERT_FILE
    [default_cert_dir] => c:/usr/local/ssl/certs
    [default_cert_dir_env] => SSL_CERT_DIR
    [default_private_dir] => c:/usr/local/ssl/private
    [default_default_cert_area] => c:/usr/local/ssl
    [ini_cafile] =>
    [ini_capath] =>
)
```

I have no drive-letter assigned to `D:` on this system, so that explains the error triggered by the first DLL, which is trying to locate a file on a drive that doesn't exist.

This also confirms my suspicion about the result being cached somehow, as, once again, the error is now no longer repeatable again - it seems to know now where to look and where not to look.

I checked the version numbers, and both DLLs are 1.0.1.16 - so it's the same tagged release, but two different builds with different configurations.

Of course, this problem could occur even with the official build, on a system that doesn't have a drive-letter `C:` - so if I had to point at a bug, I'd say the library needs to do better when checking for the presence of various files; on Windows, it would need to check for the existence of a drive-letter before checking for the presence of a file.

I don't know to whom or where to report this issue...?

**Comment by jakoch** [comment]
> this time it happens with php-win.exe only, which I discovered only because the WPN-XM control panel appears to run it on startup (?)

Yes! `php.exe` (not `php-win.exe`) is executed during "Start" to run `php -v` for picking up the version number. This check ensures that the version display of the control panel is in sync with the PHP version used. This is needed, because a user might have switched the PHP version manually between stop and start operations.

> here's the output with the DLL that comes with WPN-XM

Nope, that first output doesn't match the OpenSSL version shipped with that installer package.
The OpenSSL version packaged with `full-next-php5.6-w64`  is `openssl-x64 1.0.2c`.
Referencing: https://github.com/WPN-XM/registry/blob/master/full-0.8.6-php5.6-w64.json#L16

You wrote about swapping things out: do you mean a `OpenSSL 1.0.1.p` installation on top of WPN-XM 0.8.6? That would explain at least the version number on the path `D:\tmp\openssl-1.0.1p\vc11\x64/cert.pem`, but not that a configuration directive was set to a temporary folder. Setting config directives (especially for certificate pickups) to `tmp` is rubbish.

- Installer Check in regard to OpenSSL directives

The installer sets only the `curl.cainfo`and `openssl.capath` directives in `php.ini` during the OpenSSL installation. See: https://github.com/WPN-XM/WPN-XM/blob/master/installers/full-php56-w64.iss#L1387-L1391

It doesn't set set the environment constant ` OPENSSL_CONF` during install !
And i think it shouldn't. Because any prior installation of OpenSSL, which relies on this setting should remain intact, after installing the server stack.

- Registry Check 

The `OpenSSL 1.0.1p` version was not installed via a WPN-XM packaged installer nor the webinstaller. The version isn't in our registry, yet. 
Because: Checking the registry reveals, that i'm not even able to ship that version at all, due to abug in the OpennSLL version crawler. The version pickup failed after `1.0.1l` and starts again at `2.0.0`, see https://github.com/WPN-XM/registry/blob/master/wpnxm-software-registry.php#L1015
Due to the crawling bug the following versions are not (yet) in the registry (`1.0.1` + `m`, `o`, `p`), see https://indy.fulgan.com/SSL/

One more thing: You will find `libeay32.dll` very often on a system. Could this be a DLL shadowing issue influenced by the OpenSSL constant or the path?

- I don't know to whom or where to report this issue...?

Well, reporting to me is a start, but ... i still don't know how to reproduce it, in order to qualify it as an WPN-XM bug / configuration issue. This needs a sandbox installation - to exclude all other issues and focus on WPN-XM.

(Sidenote: I've already pointed out how WPN-XM gets to a better debugging environment. The first step is to get the "vargrant windows sandbox" up: to test the installer in an sandboxed environment. The second step is reduce the number of manual testing steps by using the task-runner combined with some "environment checking" tasks.)

Happy hacking, Jens

**Comment by mindplay-dk** [comment]
DLL shadowing, hmm, I did have some trouble with Git this week, and had to reinstall it several times - at one point, I discovered that Git's global config was pointing to a previous, non-existent path for root certificates. Could this be related?

**Comment by jakoch** [comment]
The Git issue seems to be an unrelated/independent Git configuration issue.
Its possible that the global config or the repo config contains a wrong path to the SSL certs.
And i'm again out of the game, because at the moment WPN-XM doesn't do any git configuration; except shipping a custom bash_profile.

With DLL shadowing i meant, that there is a certain search order for dynamic linked library loading.
https://msdn.microsoft.com/en-us/library/ms682586(VS.85).aspx
What happens, if libeay32.dll is already loaded in memory. Will PHP still look at the one in its path or reuse the loaded one? Its possible, that this causes trouble, but its only guessing around. 

In regard to the original issue, i still think they used a hardcoded path during compilation, which acts as default fallback. The origin of the problem is somewhere between "winlibs/openssl",  "phpext_curl" and "winlibs/curl". I guess Anatol (ab / weltling) is the only guy who could bring some light into this issue.

**Comment by mindplay-dk** [comment]
> With DLL shadowing i meant, that there is a certain search order for dynamic linked library loading

I understood that - what I was suggesting, is that maybe the Git installer did something in the registry to affect the DLL search; it's possible it's finding a different DLL now, one installed by the Git installer.

But yeah, I also have no real idea how to diagnose this any further...

**Comment by jakoch** [comment]
(For WPN-XM the following line should solve he certificate problem with Git:
`git config --system http.sslcainfo "C:\server\bin\openssl\ca-bundle.crt"`)


> maybe the Git installer did something in the registry to affect the DLL search; it's possible it's finding a different DLL now, one installed by the Git installer

Yes, it could be. One would have to check it with Process Monitor or ListDlls from Sysinternals. 

They do a lot of stuff during installation. I didn't test the latest version, yet, but i guess the wild ride continues. Now the portable package has a post-install.bat... I like the Git functionality the packages provides, but i don't like the way they install this software.

I have a different position on this, because you can't build a relocatable stack, if applications rely heavily on configuration data stored in the Windows registry. I favor simple configuration files or configuration file hierarchies. (Newer versions of Git-for-Windows use the registry key `HKEY_LOCAL_MACHINE\Software\GitForWindows` to store additional configuration data.)

**Comment by jakoch** [comment]
I hope this issue vanished over time :)

Status: I could not reproduce the issue. 
Steps explained in https://github.com/WPN-XM/WPN-XM/issues/514#issuecomment-160924577 / https://github.com/WPN-XM/WPN-XM/issues/514#issuecomment-160950331
Closing.

**Comment by mindplay-dk** [comment]
It's one of those long-standing "comes and goes" issues under Windows, but yeah, I haven't run into it recently :-)
